### PR TITLE
Support unicode escape characters

### LINKED ISSUE #638: String escape codes: should support Unicode escape codes
Similar to Javascript, I think Soldity should support Unicode escape codes in strings: `\uUUUU`, where `UUUU` is the Unicode codepoint and the result is UTF8 encoded.

### PR DESCRIPTION
**Pull request by axic**
Fixes #638

### CODE REVIEW

**Comment by chriseth** [comment]
Looks good at first glance. Thanks! Please correct compiler error and add tests.

File: libsolidity/parsing/Scanner.cpp [hunk 1]
```diff
@@ -607,10 +642,19 @@ bool Scanner::scanEscape()
 	case 'v':
 		c = '\v';
 		break;
+	case 'u':
+	{
+		unsigned codepoint;
+		if (!scanUnicode(&codepoint))
+			return false;
+		addUnicodeChar(c);
```
Comment (graup) [inline libsolidity/parsing/Scanner.cpp]: This is a bit confusing. Shouldn't it be `addUnicodeChar(codepoint);`? `c` just equals `'u'` at this point, doesn't it? Sorry if I simply understand the code wrong.

**Comment by axic** [comment]
@chriseth @graup seems like I've pushed the wrong version of the branch.

**Comment by chriseth** [comment]
As already discussed in other threads: I think we should retain the ability to add raw hex codes using `\xff` and only add utf8-encoded codes by `\u00ff`. The PR as it is now would also break backwards compatibilty.

File: libsolidity/parsing/Scanner.cpp [hunk 2]
```diff
@@ -177,6 +177,41 @@ bool Scanner::scanHexByte(char& o_scannedByte)
...
+		}
+		x = x * 16 + d;
+		advance();
+	}
+	o_codepoint = x;
+	return true;
+}
+
+// This supports codepoints between 0000 and FFFF.
+void Scanner::addUnicodeChar(unsigned codepoint)
```
Comment (chriseth) [inline libsolidity/parsing/Scanner.cpp]: `addUnicodeAsUTF8`?

File: test/libsolidity/SolidityEndToEndTest.cpp [hunk 3]
```diff
@@ -6837,6 +6837,22 @@ BOOST_AUTO_TEST_CASE(skip_dynamic_types_for_structs)
 	BOOST_CHECK(callContractFunction("g()") == encodeArgs(u256(2), u256(6)));
 }
 
+BOOST_AUTO_TEST_CASE(valid_unicode_codepoint)
+{
+	// For accessors, the dynamic types are already removed in the external signature itself.
```
Comment (chriseth) [inline test/libsolidity/SolidityEndToEndTest.cpp]: Please remove this commen.t

**Comment by axic** [comment]
This works now properly I believe.

(The Linux tests fails due to emscripten - part the ongoing Solidity standalone changes.)

**Comment by chriseth** [comment]
Please document this feature / bugfix.

**Comment by axic** [comment]
@chriseth added docs and rebased, but there's still a travis issue going on.

**Comment by chriseth** [comment]
It seems `eth` was not installed on travis somehow. Merging because tests on macos are green.

**Pull request merged**